flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/AnyToCharVarchar.java

```diff
@@ -0,0 +1,105 @@
+        super(
+                CastRulePredicate.builder()
+                        .predicate(
+                                (inputType, targetType) ->
+                                        !inputType.is(LogicalTypeRoot.ARRAY)
+                                                && targetType.is(LogicalTypeFamily.CHARACTER_STRING)
+                                                && !isStringType(targetType))
+                        .build());
+    }
+
+    /* Example generated code for STRING() -> VARCHAR(4) cast
+
+    isNull$0 = _myInputIsNull;
+    if (!isNull$0) {
+        if (_myInput.numChars() > 4) {
+            result$1 = (_myInput).substring(0, 4);
+        } else {
+            result$1 = _myInput;
+        }
+        isNull$0 = result$1 == null;
+    } else {
+        result$1 = org.apache.flink.table.data.binary.BinaryStringData.EMPTY_UTF8;
+    }
+
+    */
+    @Override
+    protected String generateCodeBlockInternal(
+            CodeGeneratorCastRule.Context context,
+            String inputTerm,
+            String returnVariable,
+            LogicalType inputLogicalType,
+            LogicalType targetLogicalType) {
+        final int precision = LogicalTypeChecks.getLength(targetLogicalType);
+        CastRule<?, ?> castRule =
+                CastRuleProvider.resolve(inputLogicalType, VarCharType.STRING_TYPE);
+
+        // Only used for non-Constructed types - for constructed, the trimming/padding is applied
+        // on each individual rule, i.e.: ArrayToStringCastRule
+        if (castRule instanceof ExpressionCodeGeneratorCastRule) {
+            @SuppressWarnings("rawtypes")
+            final String stringExpr =
+                    ((ExpressionCodeGeneratorCastRule) castRule)
+                            .generateExpression(
+                                    context, inputTerm, inputLogicalType, targetLogicalType);
+
+            CastRuleUtils.CodeWriter writer = new CastRuleUtils.CodeWriter();
+            if (context.legacyBehaviour() || !shouldPossiblyTrim(precision)) {
+                return writer.assignStmt(returnVariable, stringExpr).toString();
+            }
+            return writer.ifStmt(
+                            methodCall(stringExpr, "numChars") + " > " + precision,
+                            thenWriter ->
+                                    thenWriter.assignStmt(
+                                            returnVariable,
+                                            methodCall(
+                                                    "(" + stringExpr + ")",
+                                                    "substring",
+                                                    0,
+                                                    precision)),
+                            elseWriter -> elseWriter.assignStmt(returnVariable, stringExpr))
+                    .toString();
+        } else {
+            throw new IllegalStateException("This is a bug. Please file an issue.");
```

Comment:
I think here you still need to support non-`ExpressionCodeGeneratorCastRule`, for row, raw and multiset/map cast rules. For invoking "plain" `CodeGeneratorCastRule` perhaps just create a new context, where each new variable is added in `CodeWriter` here rather than on the context passed to this function

Reply:
I don't, because those constructed types have the trimming logic incorporated and shouldn't be "wrapped" by this rule. (break early for example).